rclpy/test/test_timer.py

```diff
@@ -148,3 +149,35 @@ def test_cancel_reset(period):
             node.destroy_node()
     finally:
         rclpy.shutdown(context=context)
+
+
+def test_time_until_next_call():
+    context = rclpy.context.Context()
+    rclpy.init(context=context)
+    try:
+        node = rclpy.create_node('test_time_until_next_call', context=context)
+        try:
+            executor = SingleThreadedExecutor(context=context)
+            try:
+                executor.add_node(node)
+                executor.spin_once(timeout_sec=0)
+
+                timer = node.create_timer(1, lambda: None)
+                try:
+                    assert not timer.is_canceled()
+                    executor.spin_once(0.1)
+                    assert timer.time_until_next_call() <= (1 * S_TO_NS)
+                    timer.reset()
+                    assert not timer.is_canceled()
+                    assert timer.time_until_next_call() <= (1 * S_TO_NS)
+                    timer.cancel()
+                    assert timer.is_canceled()
+                    assert timer.time_until_next_call() >= (1 * S_TO_NS)
```

Comment:
The implementation seems ok, I'm just not sure about returning max time here. It seems not so easy in python to know if `time_until_next_call()` is returning a valid but really long time or if the timer was canceled I would recommend either raising if the timer is canceled, or returning `None` such that this test becomes 

```python3
assert timer.time_until_next_call() is None
```

Reply:
@sloretz that totally makes sense, thanks 👍

Reply:
Just checking that you are suggesting like following with changing API?

```
bool time_until_next_call(int64_t nsec);
```

so that `time_until_next_call` python can return None if false?

Reply:
I was thinking 

```c++
std::optional<int64_t> time_until_next_call();
```

Which I think [pybind11 supports](https://pybind11.readthedocs.io/en/stable/advanced/cast/stl.html#c-17-library-containers) and we can use since [Rolling and Humble target c++ 17](https://www.ros.org/reps/rep-2000.html#rolling-ridley-june-2020-ongoing)

Reply:
I think it would be nice to have similar behaviors both in rclcpp and rclpy.
The problem is that returning `std::optional<int64_t>` in that case will be a breaking change.
Though this method is probably not used a lot, so maybe breaking it is not a hugh problem.

Reply:
> I think it would be nice to have similar behaviors both in rclcpp and rclpy.

I think it's fine for the client libraries to differ to take advantage of language features. In C++ the caller has access to `std::chrono::nanoseconds::max()`, so the return value is not too difficult to compare. In rclpy the caller is using Python and doesn't easily have access to `std::numeric_limits<...>::max()`, but Python's `None` fits pretty well.

>  Though this method is probably not used a lot, so maybe breaking it is not a hugh problem.

Breaking changes are fine in ROS Rolling

Reply:
> I think it's fine for the client libraries to differ to take advantage of language features. In C++ the caller has access to std::chrono::nanoseconds::max(), so the return value is not too difficult to compare. In rclpy the caller is using Python and doesn't easily have access to std::numeric_limits<...>::max(), but Python's None fits pretty well.

The rationale sounds good to me

Reply:
thanks for the informative comments. this has been addressed with `std::optional<int64_t>` in C++, and python returns None if the timer is canceled.